elide-async/src/main/java/com/yahoo/elide/async/service/thread/TableExportThread.java

```diff
@@ -36,33 +41,68 @@
  */
 @Slf4j
 public class TableExportThread implements Callable<AsyncAPIResult> {
+    private static final String FORWARD_SLASH = "/";
+    private static final String COMMA = ",";
+    private static final String DOUBLE_QUOTES = "\"";
+
     private TableExport queryObj;
     private TableExporter exporter;
     private Integer downloadRecordCount = 0;
     private ResultStorageEngine resultStorageEngine;
     private ObjectMapper mapper = new ObjectMapper();
-
-    public TableExportThread(TableExport queryObj, ResultStorageEngine resultStorageEngine, TableExporter exporter) {
+    private String baseURL;
+    private boolean skipCSVHeader;
+    private String apiVersion;
+    private User user;
+
+    public TableExportThread(TableExport queryObj, TableExporter exporter,
+            String baseURL, String apiVersion, User user,
+            ResultStorageEngine resultStorageEngine, boolean skipCSVHeader) {
         this.queryObj = queryObj;
         this.exporter = exporter;
         this.resultStorageEngine = resultStorageEngine;
+        this.baseURL = baseURL;
+        this.skipCSVHeader = skipCSVHeader;
+        this.apiVersion = apiVersion;
+        this.user = user;
     }
 
     @Override
     public AsyncAPIResult call() throws URISyntaxException, NoHttpResponseException, MalformedURLException {
-        Observable<PersistentResource> observableResults = exporter.export(queryObj);
+        Observable<PersistentResource> observableResults = exporter.export(queryObj, user, apiVersion);
         Observable<String> downloadString = processObservablePersistentResource(observableResults);
+        // Generate Header for CSV even when no records generated
+        if (downloadRecordCount == 0 && !skipCSVHeader && queryObj.getResultType() == ResultType.CSV) {
+            // TODO Test if this method can be used when data is present too,
+            // instead of relying on Json2Flat Header generation.
+            downloadString = generateCSVHeader(exporter.getParser(queryObj));
+        }
         storeResults(downloadString);
 
         TableExportResult queryResult = new TableExportResult();
         queryResult.setHttpStatus(200);
         queryResult.setCompletedOn(new Date());
-        //TODO Add URL generation logic and set
-        //queryResult.setUrl(new URL("url"));
+        queryResult.setUrl(generateDownloadURL());
         queryResult.setRecordCount(downloadRecordCount);
         return queryResult;
     }
 
+    private URL generateDownloadURL() {
+        log.debug("generateDownloadUrl");
+        String asyncQueryID = queryObj.getId();
```

Comment:
I think there are three places currently where we attempt to generate the downloadURL.  We need to consolidate this to one.